File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/genre/GenresDaoWrapper.java
package com.github.anrimian.musicplayer.data.database.dao.genre;

import static com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils.getSearchArgs;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.mapList;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.mapListNotNull;

import androidx.collection.LongSparseArray;
import androidx.sqlite.db.SimpleSQLiteQuery;

import com.github.anrimian.musicplayer.data.database.AppDatabase;
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao;
import com.github.anrimian.musicplayer.data.database.entities.IdPair;
import com.github.anrimian.musicplayer.data.database.entities.genres.GenreEntity;
import com.github.anrimian.musicplayer.data.database.entities.genres.GenreEntryEntity;
import com.github.anrimian.musicplayer.data.models.composition.CompositionId;
import com.github.anrimian.musicplayer.data.storage.providers.genres.StorageGenre;
import com.github.anrimian.musicplayer.data.storage.providers.genres.StorageGenreItem;
import com.github.anrimian.musicplayer.data.utils.collections.AndroidCollectionUtils;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.genres.Genre;
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre;
import com.github.anrimian.musicplayer.domain.models.order.Order;
import com.github.anrimian.musicplayer.domain.utils.ListUtils;

import java.util.List;
import java.util.Set;

import io.reactivex.rxjava3.core.Observable;

public class GenresDaoWrapper {

    private final AppDatabase appDatabase;
    private final GenreDao genreDao;
    private final CompositionsDao compositionsDao;

    public GenresDaoWrapper(AppDatabase appDatabase,
                            GenreDao genreDao,
                            CompositionsDao compositionsDao) {
        this.appDatabase = appDatabase;
        this.genreDao = genreDao;
        this.compositionsDao = compositionsDao;
    }

    public void applyChanges(List<StorageGenre> addedGenres) {
        genreDao.insertAll(mapList(addedGenres, this::toEntity));
    }

    public void applyChanges(List<StorageGenreItem> addedGenres, long genreId) {
        genreDao.insertGenreEntities(mapListNotNull(addedGenres, genre -> toEntity(genre, genreId)));
    }

    public Set<String> selectAllGenreNames() {
        return ListUtils.mapToSet(genreDao.selectAllGenreNames(), name -> name);
    }

    public LongSparseArray<StorageGenreItem> selectAllAsStorageGenreItems(long genreId) {
        return AndroidCollectionUtils.mapToSparseArray(
                genreDao.selectAllAsStorageGenreItems(genreId),
                StorageGenreItem::getId);
    }

    public Observable<List<Genre>> getAllObservable(Order order, String searchText) {
        String query = "SELECT id as id," +
                "name as name, " +
                "(SELECT count() FROM genre_entries WHERE genreId = genres.id) as compositionsCount, " +
                "(SELECT sum(duration) FROM compositions WHERE compositions.id IN (SELECT audioId FROM genre_entries WHERE genreId = genres.id)) as totalDuration " +
                "FROM genres";
        query += getSearchQuery();
        query += getOrderQuery(order);
        SimpleSQLiteQuery sqlQuery = new SimpleSQLiteQuery(query, getSearchArgs(searchText, 2));
        return genreDao.getAllObservable(sqlQuery);
    }

    public Observable<Genre> getGenreObservable(long genreId) {
        return genreDao.getGenreObservable(genreId)
                .takeWhile(list -> !list.isEmpty())
                .map(list -> list.get(0));
    }

    public Observable<List<Composition>> getCompositionsInGenreObservable(long genreId, boolean useFileName) {
        String query = GenreDao.getCompositionsQuery(useFileName);
        SimpleSQLiteQuery sqlQuery = new SimpleSQLiteQuery(query, new Object[] {genreId} );
        return genreDao.getCompositionsInGenreObservable(sqlQuery);
    }

    public List<CompositionId> getCompositionsInGenre(long genreId) {
        return genreDao.getCompositionsInGenre(genreId);
    }

    public Observable<List<ShortGenre>> getShortGenresInComposition(long compositionId) {
        return genreDao.getShortGenresInComposition(compositionId);
    }

    public List<IdPair> getGenresIds() {
        return genreDao.getGenresIds();
    }

    public void updateCompositionGenre(long compositionId, String genreName) {
        appDatabase.runInTransaction(() -> {

            // 1) find new genre by name
            Long genreId = genreDao.findGenre(genreName);

            // 2) if album not exists - create album
            if (genreId == null) {
                genreId = genreDao.insert(new GenreEntity(null, genreName));//hmm, storage?
            }

            // 3) add composition to new genre
            Long[] oldGenreIds = genreDao.getGenresByCompositionId(compositionId);
            genreDao.insertGenreEntry(new GenreEntryEntity(compositionId, genreId, null));

            if (oldGenreIds != null && oldGenreIds.length > 0) {
                genreDao.removeGenreEntry(compositionId, oldGenreIds);
                genreDao.deleteEmptyGenre(oldGenreIds);
            }
        });
    }

    public void addCompositionToGenre(long compositionId, String genreName) {
        appDatabase.runInTransaction(() -> {
            Long genreId = genreDao.findGenre(genreName);

            if (genreId == null) {
                genreId = genreDao.insert(new GenreEntity(null, genreName));//hmm, storage?
            }
            genreDao.insertGenreEntry(new GenreEntryEntity(compositionId, genreId, null));
        });
    }

    public void removeCompositionFromGenre(long compositionId, long genreId) {
        appDatabase.runInTransaction(() -> {
            genreDao.removeGenreEntry(compositionId, genreId);
            genreDao.deleteEmptyGenre(genreId);
        });
    }

    public void changeCompositionGenre(long compositionId, long oldGenreId, String newGenreName) {
        appDatabase.runInTransaction(() -> {
            Long genreId = genreDao.findGenre(newGenreName);

            if (genreId == null) {
                genreId = genreDao.insert(new GenreEntity(null, newGenreName));//hmm, storage?
            }
            genreDao.insertGenreEntry(new GenreEntryEntity(compositionId, genreId, null));
            genreDao.removeGenreEntry(compositionId, oldGenreId);

            genreDao.deleteEmptyGenre(oldGenreId);
        });
    }

    public String[] getGenreNames() {
        return genreDao.getGenreNames();
    }

    public void updateGenreName(String name, long genreId) {
        genreDao.updateGenreName(name, genreId);
    }

    public void deleteGenre(long genreId) {
        genreDao.deleteGenre(genreId);
    }

    public boolean isGenreExists(String name) {
        return genreDao.isGenreExists(name);
    }

    public boolean isGenreExists(long id) {
        return genreDao.isGenreExists(id);
    }

    public String getGenreName(long genreId) {
        return genreDao.getGenreName(genreId);
    }

    private GenreEntity toEntity(StorageGenre genre) {
        return new GenreEntity(genre.getId(), genre.getName());
    }

    private GenreEntryEntity toEntity(StorageGenreItem genre, long genreId) {
        long audioId = compositionsDao.selectIdByStorageId(genre.getAudioId());
        if (audioId == 0) {
            return null;
        }
        return new GenreEntryEntity(audioId, genreId, genre.getId());
    }

    private String getOrderQuery(Order order) {
        StringBuilder orderQuery = new StringBuilder(" ORDER BY ");
        switch (order.getOrderType()) {
            case ALPHABETICAL: {
                orderQuery.append("name");
                break;
            }
            case COMPOSITION_COUNT: {
                orderQuery.append("compositionsCount");
                break;
            }
            default: throw new IllegalStateException("unknown order type" + order);
        }
        orderQuery.append(" ");
        orderQuery.append(order.isReversed()? "DESC" : "ASC");
        return orderQuery.toString();
    }

    private String getSearchQuery() {
        return " WHERE (? IS NULL OR name LIKE ?)";
    }

}

File: data/src/main/java/com/github/anrimian/musicplayer/data/database/dao/genre/GenresDaoWrapper.kt
package com.github.anrimian.musicplayer.data.database.dao.genre

import androidx.sqlite.db.SimpleSQLiteQuery
import com.github.anrimian.musicplayer.data.database.LibraryDatabase
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao
import com.github.anrimian.musicplayer.data.database.dao.genre.GenreDao.Companion.getCompositionsQuery
import com.github.anrimian.musicplayer.data.database.utils.DatabaseUtils
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.genres.Genre
import com.github.anrimian.musicplayer.domain.models.order.Order
import com.github.anrimian.musicplayer.domain.models.order.OrderType
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import java.util.Date

class GenresDaoWrapper(
    private val appDatabase: LibraryDatabase,
    private val genreDao: GenreDao,
    private val compositionsDao: CompositionsDao,
) {

    fun getAllObservable(order: Order, searchText: String?): Observable<List<Genre>> {
        var query = """
            SELECT id as id,
            name as name, 
            (SELECT count() FROM genre_entries WHERE genreId = genres.id) as compositionsCount, 
            (SELECT sum(duration) FROM compositions WHERE compositions.id IN (SELECT compositionId FROM genre_entries WHERE genreId = genres.id)) as totalDuration 
            FROM genres
        """
        query += getSearchQuery()
        query += getOrderQuery(order)
        val sqlQuery = SimpleSQLiteQuery(query, DatabaseUtils.getSearchArgs(searchText, 2))
        return genreDao.getAllObservable(sqlQuery)
    }

    fun getGenreObservable(genreId: Long): Observable<Genre> {
        return genreDao.getGenreObservable(genreId)
            .takeWhile(List<Genre>::isNotEmpty)
            .map { list -> list[0] }
    }

    fun getCompositionsInGenreObservable(
        genreId: Long,
        useFileName: Boolean,
    ): Observable<List<Composition>> {
        val query = getCompositionsQuery(useFileName)
        val sqlQuery = SimpleSQLiteQuery(query, arrayOf(genreId))
        return genreDao.getCompositionsInGenreObservable(sqlQuery)
    }

    fun getCompositionsInGenre(genreId: Long, useFileName: Boolean): List<Composition> {
        val query: String = getCompositionsQuery(useFileName)
        val sqlQuery = SimpleSQLiteQuery(query, arrayOf(genreId))
        return genreDao.getCompositionsInGenre(sqlQuery)
    }

    fun getAllCompositionIdsByGenre(genreId: Long): Single<List<Long>> {
        return genreDao.getAllCompositionsByGenre(genreId)
    }

    fun containsCompositionGenre(compositionId: Long, genreName: String): Boolean {
        return genreDao.containsCompositionGenre(compositionId, genreName)
    }

    fun moveGenres(compositionId: Long, fromPos: Int, toPos: Int) {
        appDatabase.runInTransaction {
            genreDao.moveGenres(compositionId, fromPos, toPos)
            compositionsDao.setUpdateTime(compositionId, Date())
        }
    }

    @JvmOverloads
    fun addCompositionToGenre(compositionId: Long, genreName: String, position: Int? = null) {
        appDatabase.runInTransaction {
            var genreId = genreDao.findGenre(genreName)
            if (genreId == null) {
                genreId = genreDao.insertGenre(genreName)
            }
            if (position != null) {
                genreDao.increasePositionsAfter(position, compositionId)
            }
            genreDao.insertGenreEntry(compositionId, genreId, position)
            compositionsDao.setUpdateTime(compositionId, Date())
        }
    }

    fun removeCompositionFromGenre(compositionId: Long, genre: String): Int {
        return appDatabase.runInTransaction<Int> {
            val genreId = genreDao.findGenre(genre) ?: throw IllegalStateException("genre not found")
            val position = genreDao.getGenrePosition(compositionId, genreId)

            genreDao.removeGenreEntry(compositionId, genreId)
            genreDao.decreasePositionsAfter(position, compositionId)
            genreDao.deleteEmptyGenre(genreId)
            compositionsDao.setUpdateTime(compositionId, Date())
            return@runInTransaction position
        }
    }

    fun changeCompositionGenre(compositionId: Long, oldGenreName: String, newGenreName: String) {
        appDatabase.runInTransaction {
            val oldGenreId = genreDao.findGenre(oldGenreName)
                ?: throw IllegalStateException("old genre not found")
            val genreId = genreDao.findGenre(newGenreName)
                ?: genreDao.insertGenre(newGenreName)
            val position = genreDao.getGenrePosition(compositionId, oldGenreId)
            genreDao.insertGenreEntry(compositionId, genreId, position)
            genreDao.removeGenreEntry(compositionId, oldGenreId)
            genreDao.deleteEmptyGenre(oldGenreId)
            compositionsDao.setUpdateTime(compositionId, Date())
        }
    }

    fun getGenreNames(forCompositionId: Long): Array<String> {
        return genreDao.getGenreNames(forCompositionId)
    }

    fun updateGenreName(name: String, genreId: Long, compositionIds: List<Long>) {
        appDatabase.runInTransaction {
            genreDao.updateGenreCompositionsModifyTime(genreId, Date())
            val existsGenreId = genreDao.findGenre(name)
            if (existsGenreId == null) {
                genreDao.updateGenreName(name, genreId)
                return@runInTransaction
            }
            genreDao.changeCompositionsGenre(genreId, existsGenreId)
            genreDao.deleteEmptyGenre(genreId)
            val date = Date()
            for (compositionId in compositionIds) {
                compositionsDao.setUpdateTime(compositionId, date)
            }
        }
    }

    fun getGenreName(genreId: Long): String {
        return genreDao.getGenreName(genreId)
    }

    private fun getOrderQuery(order: Order): String {
        val orderQuery = StringBuilder(" ORDER BY ")
        when (order.orderType) {
            OrderType.NAME -> orderQuery.append("name")
            OrderType.COMPOSITION_COUNT -> orderQuery.append("compositionsCount")
            else -> throw IllegalStateException("unknown order type$order")
        }
        orderQuery.append(" ")
        orderQuery.append(if (order.isReversed) "DESC" else "ASC")
        return orderQuery.toString()
    }

    private fun getSearchQuery() = " WHERE (? IS NULL OR name LIKE ?)"
}